In [54]:
MASK = 0xFFFFFFFF
MAX_POSITIVE_INT = 0x7FFFFFFF
a = 0x80000000
print(a) # 2147483648 --> This is the minimum negative value for a 32-bit signed integer. In Python, it is treated as a normal positive number because Python integers are not fixed-width.
print(~(a ^ MASK)) # Python integers are not fixed 32-bit signed integers. They’re arbitrary precision, so 0x80000000 is interpreted as a normal positive number unless we manually simulate 32-bit overflow.
# Output will show without sign. Because Python does NOT use 32-bit signed integers. 
print(MAX_POSITIVE_INT) # 2147483647 --> This is the maximum positive value for a 32-bit signed integer(-2147483648 to 2147483647).
print(MASK) # 4294967295 --> This is the maximum value for a 32-bit unsigned integer. It is also used to mask out the sign bit when performing bitwise operations on signed integers(0 to 4294967295).

2147483648
-2147483648
2147483647
4294967295


In [28]:
print(bin(5)) # 0b101 --> 0b is a prefix that indicates that the number is in binary format. To get just the binary digits, we can slice the string starting from index 2.
print(bin(5)[2:])
# Because bin() converts an integer into its binary string representation. It’s meant for display purposes, not for numeric computation.
# If you want to perform bitwise operations, you should work with integers directly, not their string representations.
# Beacuse integer is stored as BINARY in memory, you can perform bitwise operations on them without needing to convert them to binary strings.

0b101
101


In [29]:
# Convert Binary to Decimal
binary_5 = '101'
print(int(binary_5, 2))

# Convert Binary to Hexadecimal
print(hex(int(binary_5, 2))) # 0x5 --> 0x is a prefix that indicates that the number is in hexadecimal format. To get just the hexadecimal digits, we can slice the string starting from index 2.

5
0x5


In [30]:
x = bin(5)[2:]
count = 0
for i in x:
    if i == '1':
        count += 1
print(count) # 2 --> This is the number of 1s in the binary representation of 5.

2


In [31]:
# Bitwise AND
# Truth Table for Bitwise AND
# A B   A & B 
# 0 0      0
# 0 1      0
# 1 0      0
# 1 1      1
5 & 7 # 5 in binary is 101 and 7 in binary is 111. The bitwise AND of 101 and 111 is 101, which is 5 in decimal.

5

In [32]:
# Bitwise OR
# Truth Table for Bitwise OR
# A B   A | B
# 0 0      0
# 0 1      1
# 1 0      1
# 1 1      1
5 | 7 # 5 in binary is 101 and 7 in binary is 111. The bitwise OR of 101 and 111 is 111, which is 7 in decimal.

7

In [33]:
# Bitwise XOR
# Truth Table for Bitwise XOR
# A B   A ^ B
# 0 0      0
# 0 1      1
# 1 0      1
# 1 1      0

5 ^ 7 # 5 in binary is 101 and 7 in binary is 111. The bitwise XOR of 101 and 111 is 010, which is 2 in decimal.

2

In [34]:
# Bitwise NOT
# Truth Table for Bitwise NOT
# A   ~A
# 0    1
# 1    0

~5 # 5 in binary is 101. The bitwise NOT of 101 is 010, which is 2 in decimal. However, since Python uses a signed integer representation, the result is -6. 
# This is because the bitwise NOT operation flips all the bits, including the sign bit, resulting in a negative number.

-6

In [35]:
# Left Shift
5 << 2 # 5 in binary is 101. Left shifting by 2 positions results in 10100, which is 20 in decimal.

# Right Shift
5 >> 1 # 5 in binary is 101. Right shifting by 1 position results in 10, which is 2 in decimal.

2

In [36]:
# Hexadecimal (Base 16)
print(hex(890))

3 * 16**2 + 7 * 16**1 + (10) * 16**0 # 3 * 256 + 7 * 16 + 10 * 1 = 768 + 112 + 10 = 890

0x37a


890

#### 2's complement is a way to represent negative numbers in binary. To find the 2's complement of a number, you first find the 1's complement (invert all the bits) and then add 1 to the result.
#### For example, to find the 2's complement of 5 (which is 0000 0101 in binary):
#### 1's complement of 5: 1111 1010
#### Adding 1 to the 1's complement: 1111 1010 + 1 = 1111 1011
#### The result, 1111 1011, is the 2's complement representation of -5 in binary. In decimal, this is -5.

